### ISSUE TITLE
Protect API docs behind authentication?

### ISSUE THREAD

**Issue by ghost** [tiangolo/fastapi#364]
**Basic Question**

Does FastAPI provide a method for implementing authentication middleware or similar on the docs themselves (e.g. to protect access to `/docs` and `/redoc`)?

**Additional context**

My company is currently relying on a needlessly-complex Django backend, whose main purpose is serving an API via the Django REST framework. However, it feels like half the bugs the frontend app folks are running into are caused by a disconnect between what the Django app is doing and the API documentation wiki (which is maintained by hand; it's a monster).

FastAPI thus looks like a perfect alternative. However, it's a requirement for us that the API documentation only be accessible to specific authenticated users. It looks like I could easily disable the docs when run in a production environment (as opposed to a local testing environment), but if it's possible to serve the docs on production but explicitly protect them that would be ideal (the API in part facilitates interaction between frontend apps and IOT devices, so the interactive portions of the docs would be a lot easier to use and maintain on specific hosted servers; setting up working local environments is complicated).

In my ideal world, I'd love to also auto-populate the initial authentication credentials for the interactive queries with the current user's authentication token (to allow no-configuration usage of them immediately upon access). I wasn't able to find anything in the FastAPI docs about meddling with the way the documentation is handled, but if I missed it I'd love a link!

**Comment by euri10** [comment]
You can look in the docs, there's a list of articles, one of them by niels
does that, will edit if you can't find it

**Comment by euri10** [comment]
https://medium.com/data-rebels/fastapi-how-to-add-basic-and-cookie-authentication-a45c85ef47d3

**Comment by ianbeck** [comment]
Thank you, that was exactly what I was looking for!

**Comment by tiangolo** [comment]
Thanks @euri10 for your help here! :cake: :rocket: 

Thanks @ianbeck for reporting back and closing the issue. :tada:

**Comment by ericcgu** [comment]
Put Traefik in front of it.

**Comment by imjohsep** [comment]
Medium put the article behind a paywall. Is there another source for that information?

**Comment by bastianh** [comment]
The solution discussed in the article is to disabled the automatic generated endpoints and create your own with authentication

**Comment by smentek** [comment]
> https://medium.com/data-rebels/fastapi-how-to-add-basic-and-cookie-authentication-a45c85ef47d3

I do not think it is good policy to for giving link to sources where you have to pay?! :(

**Comment by ericcgu** [comment]
use traefik forward auth

**Comment by EasonC13** [comment]
Hi there, I just find an easier solution to this question. (compare to the medium article)

You can simply disable default /docs and create docs behind authentication

Example code:

```=python
from fastapi import FastAPI

from fastapi.openapi.docs import get_swagger_ui_html

import secrets

from fastapi import Depends, FastAPI, HTTPException, status
from fastapi.security import HTTPBasic, HTTPBasicCredentials

app = FastAPI(docs_url=None, redoc_url=None)

security = HTTPBasic()


def get_current_username(credentials: HTTPBasicCredentials = Depends(security)):
    correct_username = secrets.compare_digest(credentials.username, "user")
    correct_password = secrets.compare_digest(credentials.password, "password")
    if not (correct_username and correct_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
            headers={"WWW-Authenticate": "Basic"},
        )
    return credentials.username


@app.get("/docs")
async def get_documentation(username: str = Depends(get_current_username)):
    return get_swagger_ui_html(openapi_url="/openapi.json", title="docs")
```

**Comment by kennylajara** [comment]
## Pro tips

**Using ReDoc**
If you prefer ReDoc, you can use it just replacing `get_swagger_ui_html` by `get_redoc_html`.

**Hiding those endpoints**
You are adding 2 new endpoints to your API and, almost for sure, it won't be very useful to show them in your documentation. You can hide them by including the parameter `include_in_schema=False` to the `@app.get( ... )` decorators.

**A small improvement for code mantainance**
It might be a bit better to define the title and version in the inside of `app = FastAPI( ... )` and use:
```python
return get_openapi(title=app.title, version=app.version, routes=app.routes)
```
instead of
```python
return get_openapi(title = "FastAPI", version="0.1.0", routes=app.routes)
```
This way, every-time you change the version (or if you need to change the title in the future), you can do it where you would normally do it and you won't have to remember where is that defined. Save time for the future you.

**Final code**
After does changes, this would be the final code:
```python
from fastapi import FastAPI

from fastapi.openapi.docs import get_redoc_html
from fastapi.openapi.utils import get_openapi

import secrets

from fastapi import Depends, FastAPI, HTTPException, status
from fastapi.security import HTTPBasic, HTTPBasicCredentials

app = FastAPI(
    title="FastAPI",
    version="0.1.0",
    docs_url=None,
    redoc_url=None,
    openapi_url = None,
)

security = HTTPBasic()


def get_current_username(credentials: HTTPBasicCredentials = Depends(security)):
    correct_username = secrets.compare_digest(credentials.username, "user")
    correct_password = secrets.compare_digest(credentials.password, "password")
    if not (correct_username and correct_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
            headers={"WWW-Authenticate": "Basic"},
        )
    return credentials.username


@app.get("/docs", include_in_schema=False)
async def get_documentation(username: str = Depends(get_current_username)):
    return get_redoc_html(openapi_url="/openapi.json", title="docs")


@app.get("/openapi.json", include_in_schema=False)
async def openapi(username: str = Depends(get_current_username)):
    return get_openapi(title=app.title, version=app.version, routes=app.routes
```

**Final words:** Just remember this is an example and you still need to connect this small authentication code to your real login system or, at least, use environment variables. If you are protecting your API documentation, just don't hard code your credentials...

**Comment by nissankarkifm** [comment]
> Hi there, I just find an easy solution to this question.
> 
> You can simply disable default /docs and create docs behind authentication  
  
This did not work for me. The doc page still opens up. And I can still test out the endpoints. I do see an Authorize button on the right now. Not sure if that was the intended outcome.   
How can I add a username and password prompt when trying to open the docs page?

**Comment by arshiasir** [comment]
How can this feature be implemented in other files for the router as well?

**Comment by StanlyLife** [comment]
> Medium put the article behind a paywall. Is there another source for that information?

Just delete you cookies & local storage or use incognito
